### ISSUE TITLE
git rebase upstream/master modifies third_party on the clone of forked repo

### ISSUE THREAD

**Issue by trivikr** [denoland/deno#3259]
Hi,

This is how deno workspace is set up:
```console
$ git clone --recurse-submodules https://github.com/trivikr/deno.git
$ git remote add upstream https://github.com/denoland/deno.git
$ cargo build -vv
```

After a week or two when I perform `git fetch upstream` and `git rebase upstream/master`, the following output is shown:
```console
On branch master
Your branch is ahead of 'origin/master' by 25 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   third_party (new commits)
```

Screenshot:
![Screenshot from 2019-11-03 14-09-06](https://user-images.githubusercontent.com/16024985/68092830-0f557200-fe44-11e9-8b3f-f0cb4ec3aea3.png)

How do I resolve changes in third_party?
Running `git checkout -- third_party` doesn't help

Regards,
Trivikram

**Comment by chrmoritz** [comment]
Have you tried running `git submodule update`?

**Comment by trivikr** [comment]
> Have you tried running `git submodule update`?

I just deleted my workspace, and cloned it again. So the issue is not reproducible.
I'll test running `git submodule update` when I notice this issue next time.

**Comment by trivikr** [comment]
I just ran `git submodule update` in my new cloned workspace, and the `git status` output shows unstaged commits:
```console
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)

        modified:   third_party (modified content)

no changes added to commit (use "git add" and/or "git commit -a")
```

How do I resolve these unstaged changes?

The clone is rebased with upstream:
```console
$ git remote add upstream https://github.com/denoland/deno.git
$ git fetch upstream
From https://github.com/denoland/deno
 * [new branch]        clear-cache -> upstream/clear-cache
 * [new branch]        gh-pages    -> upstream/gh-pages
 * [new branch]        golang      -> upstream/golang
 * [new branch]        master      -> upstream/master
 * [new branch]        tree-reorg  -> upstream/tree-reorg
$ git rebase upstream/master
Current branch master is up to date.
```

**Comment by chrmoritz** [comment]
Quick question: Did you touched/changed any inside `third_party` by yourself?

Whats the output of `git status` if you `cd` into `third_party`?

**Comment by trivikr** [comment]
> Did you touched/changed any inside `third_party` by yourself?

Nope, it's a fresh clone of my fork

> Whats the output of `git status` if you `cd` into `third_party`?

Attached (2.1MB file)
[third_party_git_status.txt](https://github.com/denoland/deno/files/3802240/third_party_git_status.txt)

**Comment by chrmoritz** [comment]
wait what? All these missing files. Looks like your `third_party` directory should be nearly empty. Does the build actually work that way?

Running `git reset --hard origin/HEAD` inside `third_party` should fix this for you (hopefully), but I have no idea why you ended up in this broken state.

**Comment by trivikr** [comment]
> Running `git reset --hard origin/HEAD` inside third_party should fix this for you

Yup, it does. The original question in the issue is resolved, so closing it. The conversation can continue though.

> Does the build actually work that way?

Nope, build (`cargo build`) also fails. The build succeeds after `git reset --hard origin/HEAD` in `third_party` folder

> I have no idea why you ended up in this broken state. What's your `git --version`

In the first workspace (explained in https://github.com/denoland/deno/issues/3259#issue-516897574), it happened after running `git rebase upstream/master`
In the second fresh workspace (explained in https://github.com/denoland/deno/issues/3259#issuecomment-549186091), it happened after running `git submodule update`
git version 2.17.1

**Comment by chrmoritz** [comment]
I'm glad it is working for you now.

But what you've said only explains when it happened, but not why it happened. Somehow the handling of submodules seems to be partly broken on your setup based on the 2nd attempt.
The 1st attempt is however expected behavior until you run `git submodule update`, which of course should work normally and not break the setup.

**Comment by trivikr** [comment]
Confirmed after a week that `git submodule update` fixes the original issue. It might have happened because of older version of git.

```console
$ git --version
git version 2.24.0

$ git log | head
commit b3baab6d1479abbfd2e534fe05f5e146c2d5b1a2
Author: 木杉 <zhmushan@qq.com>
Date:   Sun Nov 10 22:59:17 2019 +0800

    fix manual (#3313)
    
    PS C:\Users\zhbnh> git config --global core.symlinks=true
    error: invalid key: core.symlinks=true

commit 335e8bd33cd22891f12b58c4d9489d19c886ae4c

$ git fetch upstream
remote: Enumerating objects: 217, done.
remote: Counting objects: 100% (217/217), done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 299 (delta 197), reused 204 (delta 194), pack-reused 82
Receiving objects: 100% (299/299), 2.80 MiB | 11.20 MiB/s, done.
Resolving deltas: 100% (214/214), completed with 74 local objects.
From https://github.com/denoland/deno
   a7128baa..f001787a  gh-pages   -> upstream/gh-pages
   86b3ac51..b3baab6d  master     -> upstream/master
 * [new tag]           v0.23.0    -> v0.23.0
Fetching submodule core/libdeno/build
From https://github.com/denoland/chromium_build
 * [new branch]          20191107_win_clang_depfile -> origin/20191107_win_clang_depfile
Fetching submodule third_party
From https://github.com/denoland/deno_third_party
   3be244fe9..140d1666a  master        -> origin/master
 * [new branch]          20191107_v8   -> origin/20191107_v8
 * [new branch]          20191107_v8_2 -> origin/20191107_v8_2
 * [new branch]          20191107_v8_3 -> origin/20191107_v8_3

$ git rebase upstream/master
First, rewinding head to replay your work on top of it...
Fast-forwarded master to upstream/master.

$ git status
On branch master
Your branch is ahead of 'origin/master' by 27 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   core/libdeno/build (new commits)
        modified:   third_party (new commits)

no changes added to commit (use "git add" and/or "git commit -a")

$ git submodule update
Submodule path 'core/libdeno/build': checked out 'b3fe5cb48e9399765ec103ca0c14057f77cf9fc3'
Submodule path 'third_party': checked out '140d1666ad04fb5ddfd5fc0d55181a9c7c067327'

$ git status
On branch master
Your branch is ahead of 'origin/master' by 27 commits.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
```
